packages/network/src/workers/syncUtils.ts

```diff
@@ -485,21 +491,58 @@ export function createTransformWorldEventsFromStream(decode: ReturnType<typeof c
   };
 }
 
+function mapTransactionHashesToNetworkComponentUpdates(events: NetworkComponentUpdate[]) {
+  return events.reduce((acc, event) => {
+    if (["worker", "cache"].includes(event.txHash)) return acc;
+
+    if (!acc[event.txHash]) acc[event.txHash] = [];
+
+    acc[event.txHash].push(event);
+
+    return acc;
+  }, {} as { [key: string]: NetworkComponentUpdate[] });
+}
+
+function parseSystemCallTransactionFromStreamNetworkComponentUpdate(event: NetworkComponentUpdate) {
+  if (!event.txMetadata) return null;
+
+  const { to, data, value } = event.txMetadata;
+
+  return {
+    to,
+    data: BigNumber.from(data).toHexString(),
+    value: BigNumber.from(value),
+    hash: event.txHash,
+  } as SystemCallTransaction;
+}
+
+export function parseSystemCallsFromStreamEvents(events: NetworkComponentUpdate[]) {
+  const systemCalls: SystemCall[] = [];
+  const transactionHashToEvents = mapTransactionHashesToNetworkComponentUpdates(events);
```

Comment:
```suggestion
  const transactionHashToEvents = groupByTxHash(events);
```
(if above suggestion is accepted)